### DISCUSSION TITLE
Multilabel textcat (transformers): Weights for classes in loss function (Unbalanced data)

### DISCUSSION THREAD

**Discussion by hp52** [explosion/spaCy#7256] [Help: Coding & Implementations] [answered]
Hi,

I'm training a multilabel text classificator with transformers and Spacy 3.0. My data is quite unbalanced, so I'd like to work with loss weights for the different classes. Is it possible to define custom weights that will be used in the loss function?

Thanks for your help.

**Comment by svlandeg** [comment, marked as answer]
I think right now, the only way to accomplish that, is to create a new subclass of the `MultiLabel_TextCategorizer` object and reimplement the `get_loss` function (note that this function is defined by the class `TextCategorizer`)

**Reply by hp52** [reply]
Thank you for your answer. I'm training the classifier with the train command and config file. How would I do it in that case? Could I use the --code parameter?

**Reply by adrianeboyd** [reply]
Yes, the `--code` parameter is the right way to do this with `spacy train`.

**Reply by svlandeg** [reply]
Yes, exactly, see here: https://spacy.io/usage/training#custom-code
You'd basically call this component slightly different and then you can refer to that custom name in the config, and use `--code` to ensure that that custom name is resolved correctly.

**Reply by einarbmag** [reply]
I'm trying to figure this out as well, as I have a very highly imbalanced multiclass dataset. Using simple models in sklearn, I can confirm that weighting the classes by using `class_weights="balanced"`, I get a much better performance for the minority classes.

Looking at how to modify `get_loss` in TextCategorizer, I can see that it returns both an array of residuals, `d_scores`, and a `loss` value (MSE calculated from `d_scores`). It looks to me like `loss` is not really used to calculate parameter updates, only `d_scores`. So it seems clear that I need to weight by multiplying each element in `d_scores` with the appropriate weight (which then feed into a modified `loss` value of course, but this seems only to be used for tracking the training process). 

Is my understanding correct?

**Reply by koaning** [reply]
While it depends a bit on the task, I suppose the best thing to do is to try and collect more data from the under-represented class. In my experience, assuming that it's an option, that leads to a more clean solution than many of the over/under-sampling/weighting tricks. I usually get there by training a classifier to predict the minority class and to give those examples priority in my label queue. 

I should admit that I've never tried implementing weighting with textcat though, so I can't confirm the details.

**Reply by koaning** [reply]
One final thing I suppose I could add: a while ago I wrote a plugin, tokenwiser, that allows pre-trained scikit-learn models to act as a classifier for spaCy. Should that be of interest, the API docs can be found [here](https://koaning.github.io/tokenwiser/api/component.html).

**Reply by koaning** [reply]
Out of curiosity, is there a particular reason why you'd like to use spaCy as opposed to scikit-learn for your classification task?

**Reply by vsocrates** [reply]
@einarbmag Were you able to get this to work? I've seen some other threads saying that the `get_loss` function is often tightly coupled with the rest of the implementation but doesn't seem that way at first glance. My plan was to simply compute the weighted MSE and multiple `d_scores` by the respective weights, as you describe.

 I'd really appreciate seeing some sample could if you got it to work and would be willing to share! Thanks in advance!

**Reply by vsocrates** [reply]
I implemented by subclassing `TextCategorizer` as suggested above and overriding the relevant lines in `get_loss` method as shown below and it seems to work fine, just in case this helps anyone else! 

```python
    if weights is not None:
        squared = d_scores**2
        mean_square_error = numpy.sum(squared * weights) / numpy.sum(weights)
        d_scores *= weights
    else:
        mean_square_error = (d_scores**2).mean()
    # print(d_scores)    
    return float(mean_square_error), d_scores
```

**Reply by biomed-model** [reply]
@vsocrates I'm trying to train binary classifiers (regarding the described experimental model, e.g. animal model, cell culture, ...) using spacy.TextCatBOW.v3 for quite unbalanced classes (15% True to 85% False) in a biomedical dataset (PubMed-Abstracts). Can you provide some more information on your solution (subclassing TextCategorizer and overriding get_loss)? I'd really appreciate some guidance on this.

**Reply by vsocrates** [reply]
Here's the full file with line linked: https://github.com/vsocrates/moore-followup/blob/cfd52278bba001931a73e100eb531b4cd1dcb945/Weighted_TextCategorizer.py#L348

Hope that helps!

**Reply by biomed-model** [reply]
Yes, that helps a lot. Thank you very much!
